### ISSUE TITLE
Cannot deregister a service which was registered via Terraform without using Terraform

### ISSUE THREAD

**Issue by thaniri** [hashicorp/consul#12251]
When filing a bug, please include the following headings if possible. Any example text in this template can be deleted.

#### Overview of the Issue

When a service is registered via Terraform, it appears to be impossible to remove said service without also using Terraform. This is a problem for when a state file has been lost 🙃

#### Reproduction Steps

1. Register a service via Terraform
2. Try to deregister the service using `consul services deregister` or to remove the node running the service using `consul force-leave`

### Consul info for both Client and Server

The client doesn't exist in my case, the service just exists without an actual running server running it.

The server is running version 1.7.1.

### Operating system and Environment details

Using CentOS7 on Amd64 hardware.

### Log Fragments

Nothing gets printed when trying to run my example commands.

**Comment by jkirschner-hashicorp** [comment]
Hi @thaniri,

Thanks for reaching out. To better understand what might be happening here, can you clarify what you mean when you say:
> The client doesn't exist in my case, the service just exists without an actual running server running it.

Where were you running `consul services deregister` or `consul force-leave`? Were you calling these commands against the `consul` binary without the consul agent running on that machine?

**Comment by blake** [comment]
Hi @thaniri,

The [`consul_service`](https://registry.terraform.io/providers/hashicorp/consul/latest/docs/resources/service) Terraform resource registers services directly into Consul's catalog using the [`/v1/catalog/register`](https://www.consul.io/api/catalog#register-entity) API. When you run `consul services deregister`, that command attempts to de-register the service using the [`/agent/service/deregister/:service_id`](https://www.consul.io/api-docs/agent/service#deregister-service) endpoint.

The `/agent` APIs, however, only act on resources that are _directly registered_ to the agent which you are communicating. (See [this StackOverflow post](https://stackoverflow.com/a/65725360/12384224) for a short description of the differences between the catalog and APIs). The de-registration request is failing because the service is not registered with the agent you are communicating (I assume this is the server), but rather–as a you mentioned–a non-existent [node](https://registry.terraform.io/providers/hashicorp/consul/latest/docs/resources/service#node) that you specified in the `consul_service` resource.

Similarly, `consul force-leave <node>` attempts to mark a _real_ node as left in the gossip pools, and won't work for non-existent nodes.

The correct way to remove the node/service in this scenario is to use the [`/v1/catalog/deregister`](https://www.consul.io/api/catalog#deregister-entity) API.

For example:

```shell-session
# De-register a single service on a non-existent node
$ curl --request PUT \
       --url $CONSUL_HTTP_ADDR/v1/catalog/deregister \
       --data '{"Node": "example_node", "ServiceID": "example_service_ID"}'

# De-register all services on a non-existent node
$ curl --request PUT \
       --url $CONSUL_HTTP_ADDR/v1/catalog/deregister \
       --data '{"Node": "example_node"}'
```

I hope this makes sense. Let me know if you have any questions.

**Comment by jkirschner-hashicorp** [comment]
@thaniri : it sounds like you did some searching through the docs and found some content related to deregistration, but didn't find what you needed for your particular situation (as @blake said, the [`/v1/catalog/deregister`](https://www.consul.io/api/catalog#deregister-entity) API). Do you have any suggestions on what we could change in the docs to help someone else in your situation find the information they need?

**Comment by thaniri** [comment]
@blake your answer is excellent and has worked for me. Thank you for providing such a detailed explanation and such a simple answer to my question.

@jkirschner-hashicorp I think this is a case of PEBKAC. The documentation I was reading was particular to the agent itself, but I didn't actually have an agent to work with. Probably with a bit more determined digging through the docs I would have found it.

Perhaps I would have clued in better if the Terraform documentation for Consul resources mentioned the API calls being made, but I also could have dug into source code.

**Comment by jkirschner-hashicorp** [comment]
@thaniri : I don't believe in PEBKAC. I'm reopening this issue as a reminder to myself to look for opportunities to make this clearer in the docs. Digging through the source code should be a last resort, not something you have to do!
